### PR TITLE
[AM335x][MCASP] Add initial support

### PR DESCRIPTION
**Pull request by ldts**
Imported changes from DRA7 and adapted to AM335x

Signed-off-by: Jorge Ramirez-Ortiz <jramirez@baylibre.com>

### CODE REVIEW

**Comment by nmenon** [comment]
I have requested for team review.

File: arch/arm/mach-omap/am335x/mcasp_am335x.c [hunk 1]
```diff
@@ -0,0 +1,1423 @@
...
+	am335x_mcasp_frame_sync(regs, stream);
+	am335x_mcasp_format_units(regs, stream);
+	am335x_mcasp_serializers(regs, stream);
+	am335x_mcasp_pinctrl(regs, stream);
+
+	return 0;
+}
+
+/* ------------------------------------------------------------------------*//**
+ * @FUNCTION		am335xxx_mcasp_main
```
Comment (mesak82) [inline arch/arm/mach-omap/am335x/mcasp_am335x.c]: s/am335xxx/am335x/g

**Review by mesak82 — COMMENTED** [review]
My only concern is that there will be duplication of the register 'dump' and 'show' command implementation for every platform where we wanted to have McASP commands supported, in spite of being the same McASP hardware module.  I assume that the decoding of the register is same as DRA7xx's except for the clocking checks.

File: arch/arm/mach-omap/am335x/mcasp_am335x.c (hunk 1 above)
Comment (nmenon) [reply arch/arm/mach-omap/am335x/mcasp_am335x.c]: >there will be duplication of the register 'dump' and 'show' command 
Yes, unfortunately so, but that is something of the original architectural limitation.

File: arch/arm/mach-omap/am335x/mcasp_am335x.h [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+#include <stdio.h>
+
+#define MCASP1			1
+#define MCASP2			2
+
+#define MCASP1_CFG_BASEADDR	0x48038000
+#define MCASP2_CFG_BASEADDR	0x4803C000
+
+#define MCASP_PID_OFFSET	0x00
+#define PWRIDLESYSCONFIG_OFFSET	0x04
```
Comment (dgerlach) [inline arch/arm/mach-omap/am335x/mcasp_am335x.h]: Why is this one register named differently? Why not just MCASP_SYSCONFIG_OFFSET?

Comment (mesak82) [reply arch/arm/mach-omap/am335x/mcasp_am335x.h]: That's the register name as per TRM. It's the only register whose name is not prefixed with "MCASP_".

**Comment by dgerlach** [comment]
What is the expected output of `omapconf show mcasp` or 'omapconf dump mcasp`? Which hardware platform and which Linux kernel was this tested against? I see no output when using AM335x EVM and TI kernel.

**Comment by mesak82** [comment]
@dgerlach - the command must have a mcasp instance number, i.e. "omapconf show mcasp1". If McASPx is not active at the time, it should print a message saying that the module is not clocked (or something like that). Otherwise, when McASPx is active (i.e. audio playing), "omapconf show mcaspX" should print several tables with the current state of McASP (frame sync generator, serializer, FIFOs, etc).

**Comment by ldts** [comment]
@dgerlach @mesak82 https://hastebin.com/uhukiralem.rb

**Comment by dgerlach** [comment]
Please add help for this command -> arch/arm/mach-omap/am335x/help_am335x.c

**Comment by dgerlach** [comment]
Also @mesak82 thanks for the description, currently `omapconf dump mcasp` fails silently which is probably not what we want and confusing.

**Comment by dgerlach** [comment]
Can confirm that if I play audio and invoke `omapconf show mcasp2` and `omapconf dump mcasp2` I see proper output.

File: arch/arm/mach-omap/am335x/mcasp_am335x.h [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+#define MCASP_TXTDMSLOT_OFFSET	0xc4
+#define MCASP_TXCLKCHK_OFFSET	0xc8
+#define MCASP_XEVTCTL_OFFSET	0xcc
+
+#define MCASP_XRSRCTL_OFFSET(n)	(0x180 + 0x04*(n))
+
+#define MCASP_WFIFOCTL_OFFSET	0x1000
+#define MCASP_WFIFOSTS_OFFSET	0x1004
+#define MCASP_RFIFOCTL_OFFSET	0x1008
+#define MCASP_RFIFOSTS_OFFSET	0x100c
```
[the remark this answers is not in the document]
Comment (nmenon) [reply arch/arm/mach-omap/am335x/mcasp_am335x.h]: unfortunately, at this time - yes.

[the remark this answers is not in the document]
Comment (ujfalusi) [reply arch/arm/mach-omap/am335x/mcasp_am335x.h]: So if we add McASP support for am437x, we duplicate the same code again?

**Pull request closed without merging**